image/postgres/scripts/init-entrypoint.sh

```diff
@@ -5,4 +5,50 @@ set -Eeo pipefail
 # Initialize DB if it does not exist
 if [ ! -s "$PGDATA/PG_VERSION" ]; then
   initdb --auth-host=scram-sha-256 --auth-local=scram-sha-256 --pwfile /run/secrets/stackrox.io/secrets/password --data-checksums
+else
+    # Verify if we need to perform major version upgrade
+    PG_BINARY_VERSION=$(postgres -V |\
+        sed 's/postgres (PostgreSQL) \([0-9]*\).\([0-9]*\).*/\1/')
+
+    if [ $(cat "${PGDATA}/PG_VERSION") -lt "$PG_BINARY_VERSION" ]; then
+        # Data version is less than binaries, upgrade
+
+        # Do a backup first, ideally move it to a separate volume. Since the
+        # database is stopped we may as well simple take a filesystem backup.
+        # Alternatives would be pg_dump or pg_basebackup, both require running
+        # database cluster.
+        echo "Backup..."
+        export BACKUP_DIR="/tmp/backups/$(date +%s)"
+        mkdir -p $BACKUP_DIR
+        tar -cf $BACKUP_DIR/backup.tar -C $PGDATA .
```

Comment:
It might also make sense to enable at least some low level of compression. Seems like we're more concerned about the disk space/throughput rather than maxing out CPU?

Reply:
Not only, it also affects the downtime -- the more compression is involved the more CPU resources we would require and the longer it will take. All in all we would need to find just a right balance.

Reply:
> the more compression is involved the more CPU resources we would require

true

> and the longer it will take

Not necessarily. Waiting on I/O might dominate the time, and decreasing the amount of written data by compressing might actually decrease the write time. CPUs are fast these days, and level 1 or 2 gzip compression is pretty fast.

Reply:
To close up on this, at least in testing on a default gke volumes compression did noticeably reduce backup time.